// chimp-widgets/lib/list.js
// Generated by CoffeeScript 1.9.2
(function() {
  var List, Widget,
    bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  Widget = require('./widget');

  List = (function(superClass) {
    extend(List, superClass);

    List.prototype.widgets = null;

    List.prototype.itemSelector = null;

    List.prototype._nestedItemsSelector = null;

    function List(selector, itemSelector, driver, widget, promise) {
      this._wrapAsWidgets = bind(this._wrapAsWidgets, this);
      List.__super__.constructor.call(this, selector, driver, promise);
      this.Widget = widget != null ? widget : Widget;
      if (this.itemSelector == null) {
        this.itemSelector = itemSelector;
      }
      this._nestedItemsSelector = this.selector + " " + this.itemSelector;
      this.widgets = new this.Promise((function(_this) {
        return function(fulfill, reject) {
          return _this.driver.elements(_this._nestedItemsSelector, function(error, elements) {
            if (error != null) {
              return reject(error);
            } else {
              return fulfill(_this._wrapAsWidgets(elements.value));
            }
          });
        };
      })(this));
    }

    List.prototype.findWhere = function(filter) {
      return this.widgets.then((function(_this) {
        return function(widgets) {
          return _this.Promise.any(widgets.map(function(widget) {
            return filter(widget).then(function(result) {
              if (result) {
                return _this.Promise.resolve(widget);
              } else {
                return _this.Promise.reject();
              }
            });
          }));
        };
      })(this));
    };

    List.prototype.findByText = function(text) {
      return this.findWhere((function(_this) {
        return function(widget) {
          return widget.hasText(text).then(function() {
            return widget;
          });
        };
      })(this));
    };

    List.prototype.map = function(mapper) {
      return this.widgets.then((function(_this) {
        return function(widgets) {
          return widgets.map(mapper);
        };
      })(this));
    };

    List.prototype._wrapAsWidgets = function(elements) {
      return elements.map((function(_this) {
        return function(element, index) {
          return new _this.Widget(_this._nestedItemsSelector + ":nth-child(" + (index + 1) + ")");
        };
      })(this));
    };

    return List;

  })(Widget);

  module.exports = List;

}).call(this);
